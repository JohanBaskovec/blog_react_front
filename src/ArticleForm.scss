.ArticleForm {
  display: flex;
  flex-grow: 1;
  flex-direction: column;
}
.ArticleForm__buttons-container {
  margin-bottom: 2rem;
}
