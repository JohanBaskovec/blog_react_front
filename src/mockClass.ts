

type constructorTypeOf<T> = new (...args:any[]) => T;

export function mockDefaultApi<T>(o: T): T {
    const mocked: any = {};
    for (const key in o) {
        if (typeof o[key] === 'function') {
            mocked[key] = jest.fn();
        }
    }
    return mocked;
}

type MockedProxy<T> = jest.Mock<T extends (...args: any[]) => infer U ? U : T, any[]>;

type MockedClass<T> = {
    [k in keyof T]: MockedProxy<T[k]>;
}

export function mockClass<T>(constr: constructorTypeOf<T>): T & MockedClass<T>{
    let obj: T = new constr();
    const mocked: any = {};

    do {
        const properties = Object.getOwnPropertyNames(obj);
        for (const propertyName of properties) {
            if (typeof (obj as any)[propertyName] === "function" && !mocked[propertyName]) {
                mocked[propertyName] = jest.fn();
            }
        }
    } while (obj = Object.getPrototypeOf(obj));
    return mocked as any;
}
