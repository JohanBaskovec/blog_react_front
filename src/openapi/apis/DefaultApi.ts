// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { Observable } from 'rxjs';
import { BaseAPI, HttpHeaders, throwIfNullOrUndefined, encodeURI } from '../runtime';
import {
    Article,
    ArticleFormData,
    LoginForm,
    RegistrationForm,
    User,
} from '../models';

export interface GetArticleByIdRequest {
    id: string;
}

export interface InsertArticleRequest {
    articleFormData: ArticleFormData;
}

export interface LoginRequest {
    loginForm: LoginForm;
}

export interface RegisterRequest {
    registrationForm: RegistrationForm;
}

export interface UpdateArticleRequest {
    articleFormData: ArticleFormData;
}

/**
 * no description
 */
export class DefaultApi extends BaseAPI {

    /**
     * Get all the articles
     */
    getAllArticles = (): Observable<Array<Article>> => {
        return this.request<Array<Article>>({
            path: '/article',
            method: 'GET',
        });
    };

    /**
     * Get an article
     */
    getArticleById = ({ id }: GetArticleByIdRequest): Observable<Article> => {
        throwIfNullOrUndefined(id, 'getArticleById');

        return this.request<Article>({
            path: '/article/{id}'.replace('{id}', encodeURI(id)),
            method: 'GET',
        });
    };

    /**
     * Get the current session
     */
    getCurrentAuthenticatedUser = (): Observable<User> => {
        const headers: HttpHeaders = {
        };

        return this.request<User>({
            path: '/me',
            method: 'GET',
            headers,
        });
    };

    /**
     * Insert an article
     */
    insertArticle = ({ articleFormData }: InsertArticleRequest): Observable<void> => {
        throwIfNullOrUndefined(articleFormData, 'insertArticle');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/article',
            method: 'POST',
            headers,
            body: articleFormData,
        });
    };

    /**
     * Create a new HTTP session
     */
    login = ({ loginForm }: LoginRequest): Observable<User> => {
        throwIfNullOrUndefined(loginForm, 'login');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<User>({
            path: '/login',
            method: 'POST',
            headers,
            body: loginForm,
        });
    };

    /**
     * Logout
     */
    logout = (): Observable<void> => {
        const headers: HttpHeaders = {
        };

        return this.request<void>({
            path: '/logout',
            method: 'POST',
            headers,
        });
    };

    /**
     * Register
     */
    register = ({ registrationForm }: RegisterRequest): Observable<void> => {
        throwIfNullOrUndefined(registrationForm, 'register');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/register',
            method: 'POST',
            headers,
            body: registrationForm,
        });
    };

    /**
     * Update an article
     */
    updateArticle = ({ articleFormData }: UpdateArticleRequest): Observable<void> => {
        throwIfNullOrUndefined(articleFormData, 'updateArticle');

        const headers: HttpHeaders = {
            'Content-Type': 'application/json',
        };

        return this.request<void>({
            path: '/article',
            method: 'PUT',
            headers,
            body: articleFormData,
        });
    };

}
