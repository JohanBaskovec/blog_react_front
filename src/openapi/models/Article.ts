// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * @export
 * @interface Article
 */
export interface Article {
    /**
     * @type {string}
     * @memberof Article
     */
    id: string;
    /**
     * @type {string}
     * @memberof Article
     */
    title: string;
    /**
     * @type {string}
     * @memberof Article
     */
    content: string;
}
