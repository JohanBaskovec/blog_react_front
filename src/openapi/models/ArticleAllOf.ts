// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import {
    User,
} from './';

/**
 * @export
 * @interface ArticleAllOf
 */
export interface ArticleAllOf {
    /**
     * @type {string}
     * @memberof ArticleAllOf
     */
    title: string;
    /**
     * @type {string}
     * @memberof ArticleAllOf
     */
    content: string;
    /**
     * @type {User}
     * @memberof ArticleAllOf
     */
    author: User;
    /**
     * @type {number}
     * @memberof ArticleAllOf
     */
    version: number;
}
