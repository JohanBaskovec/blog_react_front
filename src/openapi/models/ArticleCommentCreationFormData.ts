// tslint:disable
/**
 * Blog
 * No description provided (generated by Openapi Generator https://github.com/openapitools/openapi-generator)
 *
 * The version of the OpenAPI document: 0.0.0
 * Contact: johan@johanbaskovec.fr
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * @export
 * @interface ArticleCommentCreationFormData
 */
export interface ArticleCommentCreationFormData {
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    id: string;
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    articleId: string;
    /**
     * @type {string}
     * @memberof ArticleCommentCreationFormData
     */
    content: string;
}
