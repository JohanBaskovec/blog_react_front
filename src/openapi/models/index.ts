export * from './Article';
export * from './HttpSession';
export * from './LoginForm';
export * from './RegistrationForm';
export * from './User';
