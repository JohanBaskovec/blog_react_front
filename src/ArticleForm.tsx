import {DefaultApi} from "./openapi/apis";
import React, {FormEvent, useEffect, useState} from "react";
import {useParams} from "react-router-dom";
import "./ArticleForm.scss";
import {Form} from "./form/Form";
import {InputFormGroup} from "./form/InputFormGroup";
import {TextAreaFormGroup} from "./form/TextAreaFormGroup";
import {Article} from "./openapi/models";
import {ApiError} from "./ApiError";
import {ValueChangeEvent} from "./form/ValueChangeEvent";
import {Button} from "./form/Button";
import {FormButtonsContainer} from "./form/FormButtonsContainer";

interface ArticleFormParams {
    id: string;
}

interface ArticleFormProps {
    api: DefaultApi;
}

export function ArticleForm({api}: ArticleFormProps) {
    const {id}: ArticleFormParams = useParams<ArticleFormParams>();
    const [article, setArticle] = useState<Article | null>(null);
    const [error, setError] = useState<ApiError | null>(null);

    useEffect(() => {
        api.getArticleById({id: id}).subscribe((article) => {
                setArticle(article);
            },
            (error) => {
                setError(ApiError.fromError(error));
            });
    }, [])

    if (article == null) {
        return (<div>Loading...</div>);
    }

    const handleChange = (e: ValueChangeEvent<string>) => {
        setArticle({...article, [e.name]: e.value});
    }

    const submit = (e: React.MouseEvent<HTMLButtonElement>) => {
        api.updateArticle({article}).subscribe(() => {

        });
    }
    return (
        <div className="ArticleForm">
            <Form name="article-form">
                <InputFormGroup inputName={'title'}
                                inputLabel={'Title'}
                                onValueChange={handleChange}
                                value={article.title}
                                style={{marginBottom: "1rem"}}>
                </InputFormGroup>
                <TextAreaFormGroup value={article.content}
                                   onValueChange={handleChange}
                                   inputName="content"
                                   style={{marginBottom: "1rem"}}
                                   inputLabel="Content">
                </TextAreaFormGroup>
                <FormButtonsContainer>
                    <Button onClick={submit}>Submit</Button>
                </FormButtonsContainer>
            </Form>
        </div>);
}
