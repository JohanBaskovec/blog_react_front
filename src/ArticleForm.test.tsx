import React from "react";
import {DefaultApi, GetArticleByIdRequest} from "./openapi/apis";
import {of} from "rxjs";
import {mockClass} from "./mockClass";
import {render, unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import pretty from "pretty";
import {Article} from "./openapi/models";
import {MemoryRouter} from "react-router-dom";
import {ArticleForm} from "./ArticleForm";

let container: HTMLDivElement;
beforeEach(() => {
    // setup a DOM element as a render target
    container = document.createElement("div");
    document.body.appendChild(container);
});

afterEach(() => {
    // cleanup on exiting
    if (container != null) {
        unmountComponentAtNode(container);
        container.remove();
    }
});

describe('ArticleForm', () => {
    it("renders the article form when it exists", () => {
        const api: DefaultApi = mockClass(new DefaultApi());
        api.getArticleById = jest.fn(({id}: GetArticleByIdRequest) => {
            const article: Article = {
                id: id,
                title: 'title',
                content: 'content',
            }
            return of(article);
        });
        act(() => {
            render(
                <MemoryRouter initialEntries={['/article/3']}>
                    <ArticleForm api={api}/>
                </MemoryRouter>
                , container);
        });
        expect(pretty(container.innerHTML)).toMatchSnapshot();
    });
});
