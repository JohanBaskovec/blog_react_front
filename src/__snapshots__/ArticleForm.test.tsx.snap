// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleForm renders the article form when it exists 1`] = `
"<div class=\\"ArticleForm\\">
  <form class=\\"Form\\">
    <div class=\\"InputFormGroup\\" style=\\"margin-bottom: 1rem;\\"><label class=\\"Label\\" for=\\"title\\">Title</label><input name=\\"title\\" id=\\"title\\" class=\\"Input\\" value=\\"title\\"></div>
    <div class=\\"TextAreaFormGroup\\" style=\\"margin-bottom: 1rem;\\"><label class=\\"Label\\" for=\\"content\\">Content</label><textarea id=\\"content\\" class=\\"TextArea\\">content</textarea></div>
    <div class=\\"FormButtonsContainer\\"><button type=\\"button\\" class=\\"Button\\">Submit</button></div>
  </form>
</div>"
`;
